Core/Core/Dashboard/K5/ViewModel/Grades/K5GradesViewModel.swift

```diff
@@ -16,13 +16,112 @@
+                                        courseID: $0.id)
+        }
+        var gradingPeriodModels = courses.compactMap { $0.gradingPeriods }.flatMap { $0 }
+        gradingPeriodModels.sort(by: {
+            guard let date0 = $0.startDate, let date1 = $1.startDate else { return false }
+            return date0 < date1
+        })
+        gradingPeriods.append(contentsOf: gradingPeriodModels.map { return K5GradingPeriod(periodID: $0.id, title: $0.title) })
+        finishRefresh()
+    }
+
+    private func updateEnrollments(for gradingPeriodID: String) {
+        grades.removeAll()
+        let request = GetEnrollmentsRequest(context: .currentUser, userID: studentID, gradingPeriodID: gradingPeriodID, types: [ "StudentEnrollment" ], states: [ .active ])
+        env.api.makeRequest(request) { [weak self] apiEnrollments, _, _ in
+            var grades: [K5GradeCellViewModel] = []
+            apiEnrollments?.forEach { enrollment in
+                guard let course = self?.courses.first(where: { $0.id == enrollment.course_id?.rawValue }), !course.isHomeroomCourse else { return }
+                let cellModel = K5GradeCellViewModel(a11yId: "K5GradeCell.\(course.id)",
+                                                     title: course.name ?? "",
+                                                     imageURL: course.imageDownloadURL,
+                                                     grade: enrollment.computed_current_grade ?? enrollment.grades?.current_grade,
+                                                     score: enrollment.computed_current_score ?? enrollment.grades?.current_score,
+                                                     color: course.color,
+                                                     courseID: course.id)
+                grades.append(cellModel)
+            }
+            grades.sort(by: {$0.title < $1.title})
+            performUIUpdate {
+                self?.grades = grades
+            }
+        }
+        finishRefresh()
+    }
+
+    private func finishRefresh() {
+        forceRefresh = false
+        performUIUpdate {
+            self.refreshCompletion?()
+            self.refreshCompletion = nil
+        }
+    }
+
+    public func didSelect(gradingPeriod: K5GradingPeriod) {
+        currentGradingPeriod = gradingPeriod
+        reloadData()
+    }
 }
 
 extension K5GradesViewModel: Refreshable {
 
-    func refresh(completion: @escaping () -> Void) {
-        DispatchQueue.main.asyncAfter(deadline: .now() + 2, execute: completion)
+    public func refresh(completion: @escaping () -> Void) {
+        forceRefresh = true
+        refreshCompletion = completion
+        reloadData()
+    }
+
+    func reloadData() {
+        guard let periodID = currentGradingPeriod.periodID else {
+            courses.exhaust(force: true)
+            return
+        }
+        updateEnrollments(for: periodID)
+    }
+}
+
+public class K5GradingPeriod: NSObject {
```

Comment:
Why do we need `NSObject` here?

Reply:
guess it was for foreach, changed it to a hashable struct